Content.Server/EnergyDome/EnergyDomeSystem.cs

```diff
@@ -0,0 +1,259 @@
+        }
+    }
+
+    public bool AttemptUse(Entity<EnergyDomeGeneratorComponent> generator, EntityUid user)
+    {
+        if (TryComp(generator, out LockComponent? lockComp) && lockComp.Locked)
+        {
+            _audio.PlayPvs(generator.Comp.AccessDeniedSound, generator);
+            _popup.PopupEntity(
+                Loc.GetString("energy-dome-access-denied"),
+                generator,
+                user);
+            return false;
+        }
+
+        if (!_powerCell.TryGetBatteryFromSlot(generator, out var battery) && !TryComp(generator, out battery))
+        {
+            _audio.PlayPvs(generator.Comp.TurnOffSound, generator);
+            _popup.PopupEntity(
+                Loc.GetString("energy-dome-no-cell"),
+                generator,
+                user);
+            return false;
+        }
+
+        if (generator.Comp.NextActivation > _gameTiming.CurTime)
+        {
+            _audio.PlayPvs(generator.Comp.TurnOffSound, generator);
+            _popup.PopupEntity(
+                Loc.GetString("energy-dome-recharging"),
+                generator,
+                user);
+            return false;
+        }
+
+
+
+        if (battery.Charge < generator.Comp.Wattage)
+        {
+            _audio.PlayPvs(generator.Comp.TurnOffSound, generator);
+            _popup.PopupEntity(
+                Loc.GetString("energy-dome-no-power"),
+                generator,
+                user);
+            return false;
+        }
+
+        Toggle(generator);
+        return true;
+    }
+
+    public void Toggle(Entity<EnergyDomeGeneratorComponent> generator)
+    {
+        if (generator.Comp.Enabled)
+        {
+            TurnOff(generator, false);
+        }
+        else
+        {
+            TurnOn(generator);
+        }
+    }
+
+    public void TurnOn(Entity<EnergyDomeGeneratorComponent> generator)
+    {
+        if (generator.Comp.Enabled)
+            return;
+
+        generator.Comp.Enabled = true;
+        var newDome = Spawn(generator.Comp.DomePrototype, Transform(generator).Coordinates);
+
+        if (TryComp<EnergyDomeComponent>(newDome, out var domeComp))
+        {
+            domeComp.Generator = generator;
+        }
+
+        generator.Comp.SpawnedDome = newDome;
+        _audio.PlayPvs(generator.Comp.TurnOnSound, generator);
+    }
+
+    public void TurnOff(Entity<EnergyDomeGeneratorComponent> generator, bool startReloading)
+    {
+        if (!generator.Comp.Enabled)
+            return;
+
+        generator.Comp.Enabled = false;
+        QueueDel(generator.Comp.SpawnedDome);
+
+        _audio.PlayPvs(generator.Comp.TurnOffSound, generator);
+        if (startReloading)
+        {
+            generator.Comp.NextActivation = _gameTiming.CurTime + TimeSpan.FromSeconds(generator.Comp.ReloadSecond);
+            _audio.PlayPvs(generator.Comp.EnergyOutSound, generator);
+        }
+    }
+
+    public override void Update(float frameTime)
+    {
```

Comment:
A lot of this code is just duplicating existing functionality e.g.
- UseDelay
- ItemToggle
- PowerCellDraw

Your components should be re-useable where possible otherwise we have to copy-paste the same item activation code dozens of times and it becomes impossible to manage.

Reply:
Uh-oh. I was oriented to lantern this time, seems like it was a bad reference

Reply:
I added modular support for powercelldraw and usedelay, thanks for the tip! But I'm looking at ItemToggle and I VERY much don't want to use it

it seems to me that with its help it is not so convenient to realize checks of the possibility of switching the object. The system has a very sparse interface

Reply:
I don't follow? ItemToggle is for generic item turn on / turn off where systems can subscribe to events as instead we get energysword / stunbaton / powercell / welding / energy dome all copying each other with varying degrees of bugs.